### PR TITLE
8293910: tools/launcher/FXLauncherTest.java fail with jfx

### PR DESCRIPTION
**Pull request by xiangzhai**
Hi,

@dumasun reported the issue:

Configured with jfx-ls-modular-sdk:

```
configure --with-import-modules=modular-sdk
```

`make run-test CONF=fastdebug TEST="tools/launcher/FXLauncherTest.java"` failed:

```
----------System.err:(11/697)----------
java.lang.RuntimeException: JavaFX modules erroneously included in the JDK
    at FXLauncherTest.main(FXLauncherTest.java:451)
    at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
    at java.base/java.lang.reflect.Method.invoke(Method.java:578)
    at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
    at java.base/java.lang.Thread.run(Thread.java:1589)

JavaTest Message: Test threw exception: java.lang.RuntimeException: JavaFX modules erroneously included in the JDK
JavaTest Message: shutting down test

STATUS:Failed.`main' threw exception: java.lang.RuntimeException: JavaFX modules erroneously included in the JDK
```

Thanks,
Leslie Zhai

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293910](https://bugs.openjdk.org/browse/JDK-8293910): tools/launcher/FXLauncherTest.java fail with jfx

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10299/head:pull/10299` \
`$ git checkout pull/10299`

Update a local copy of the PR: \
`$ git checkout pull/10299` \
`$ git pull https://git.openjdk.org/jdk pull/10299/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10299`

View PR using the GUI difftool: \
`$ git pr show -t 10299`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10299.diff">https://git.openjdk.org/jdk/pull/10299.diff</a>

</details>

### CODE REVIEW

**Comment by xiangzhai** [comment]
Hi,

Thank @kevinrushforth [8202553: Update FXLauncherTest as part of removing JavaFX from JDK](https://github.com/openjdk/jdk/commit/8af434d695ca2f8da8fcdbbf058b85d4e79fbd1a).  And jdk11u-dev Passed:

```
make run-test CONF=fastdebug TEST="tools/launcher/FXLauncherTest.java"

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/tools/launcher/FXLauncherTest.java     1     1     0     0   
==============================
TEST SUCCESS
```

So I just skip the test for configured with jfx-ls-modular-sdk.

Please review my patch.

Thanks,
Leslie Zhai

**Comment by AlanBateman** [comment]
Can you try the following:

1. Change doFxCompile and doFxExec to use --upgrade-module-path instead of --module-path?
2. Remove the check for javafx.application.Application from the main method.

It would be interesting to check if it passes with regular builds of the JDK and your build that includes the JavaFX modules.

Note that the test summary includes the sentence "It also verifies that FX is, in fact, not included with the JDK" so that will need to change if the test is changed.

**Comment by kevinrushforth** [comment]
It does seem worth trying `--patch-module` if the javafx.graphics module is present in the Java runtime. If that works, it is better than skipping the test, since you will then be able to continue testing the FX launcher.

Another possibility would be to go with your proposed solution of skipping the test if the javafx.graphics module is present, and provide a second test that is only run when the `javafx.graphics` module is present. In that case, you could test the actual FX launcher rather than the dummy implementation. This could either be done in this PR or as a follow-on test bug.

File: test/jdk/tools/launcher/FXLauncherTest.java [hunk 1]
```diff
@@ -232,8 +242,13 @@ static void doFxCompile(String...compilerArgs) {
     static TestResult doFxExec(String...cmds) {
         List<String> fxCmds = new ArrayList<>();
         fxCmds.addAll(Arrays.asList(cmds));
-        fxCmds.add(1, "--module-path=" + MODULE_DIR);
-        fxCmds.add(2, "--add-modules=javafx.graphics");
+        if (fxClass == null) {
+            fxCmds.add(1, "--module-path=" + MODULE_DIR);
+            fxCmds.add(2, "--add-modules=javafx.graphics");
+        } else {
+            fxCmds.add(1, "--upgrade-module-path=" + IMPORT_MODULE_DIR);
```
Comment (xiangzhai) [inline test/jdk/tools/launcher/FXLauncherTest.java:249]: Hi @kevinrushforth @AlanBateman 

Please point out my fault if I misunderstanding.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
Hi @kevinrushforth @AlanBateman

I have tried `--upgrade-module-path` and `--patch-module` but still failed. Please point out my fault if I misunderstanding.

Thanks,
Leslie Zhai

**Comment by kevinrushforth** [comment]
How does the test fail? Does it actually find and load the "mockfx" classes, but subsequently fail? Or is it unable to patch/upgrade the javafx.graphics module with the "mockfx" classes?

**Comment by xiangzhai** [comment]
Hi @kevinrushforth 

I updated the testcase to check whether or not FX is part of jdk:

```
        // Check whether or not FX is part of jdk
        try {
            fxClass = Class.forName(FX_MARKER_CLASS);
        } catch (ClassNotFoundException ex) {
            // do nothing
        }
        if (fxClass != null) {
            System.out.println("JavaFX modules included in the JDK");
        }
```

And tried to `--upgrade-module-path` in the `doFxCompile` when JavaFX modules included in the JDK:

```
    static void doFxCompile(String...compilerArgs) {
        if (fxClass == null) {
            compileFXModule();
        }

        List<String> fxCompilerArgs = new ArrayList<>();
        if (fxClass == null) {
            fxCompilerArgs.add("--module-path=" + MODULE_DIR);
        } else {
            fxCompilerArgs.add("--upgrade-module-path=" + IMPORT_MODULE_DIR);
        }
        fxCompilerArgs.add("--add-modules=javafx.graphics");
        fxCompilerArgs.addAll(Arrays.asList(compilerArgs));
        compile(fxCompilerArgs.toArray(new String[fxCompilerArgs.size()]));
    }
```

Also tried `--patch-module` in the `doFxExec`:

```
    static TestResult doFxExec(String...cmds) {
        List<String> fxCmds = new ArrayList<>();
        fxCmds.addAll(Arrays.asList(cmds));
        if (fxClass == null) {
            fxCmds.add(1, "--module-path=" + MODULE_DIR);
            fxCmds.add(2, "--add-modules=javafx.graphics");
        } else {
            fxCmds.add(1, "--upgrade-module-path=" + IMPORT_MODULE_DIR);
            fxCmds.add(2, "--patch-module=javafx.graphics=" + IMPORT_MODULE_DIR + "/javafx.graphics.jmod");
        }
        return doExec(fxCmds.toArray(new String[fxCmds.size()]));
    }
```

But it might be unable to patch/upgrade the javafx.graphics module with the "mockfx" classes:

```
STDOUT:
_____________________________________________
test# 1-  Main method: public static void main(String[] args) { launch(args); }
Main-Class: helloworld.HelloWorld
renaming fxtest.jar to testBasicFXApp_useCP_noJAC-1-fxtest.jar
FAIL: testBasicFXApp_useCP_noJAC:1 : test with helloworld.HelloWorld
Test static void FXLauncherTest.testBasicFXApp() throws java.lang.Exception failed: java.lang.reflect.InvocationTargetException 
----begin detailed exceptions----
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at TestHelper.run(TestHelper.java:200)
	at FXLauncherTest.main(FXLauncherTest.java:469)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.Exception: Failed: testBasicFXApp_useCP_noJAC:1
	at FXLauncherTest.checkStatus(FXLauncherTest.java:205)
	at FXLauncherTest.testBasicFXApp(FXLauncherTest.java:311)
	at FXLauncherTest.testBasicFXApp(FXLauncherTest.java:266)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 7 more
----end detailed exceptions----
_____________________________________________
test# 2-  Main method: public static void main(String[] args) { launch(args); }
Main-Class: helloworld.ExtHello
renaming fxtest.jar to testExtendFXApp_useCP_noJAC-2-fxtest.jar
FAIL: testExtendFXApp_useCP_noJAC:2 : test with helloworld.ExtHello
Test static void FXLauncherTest.testExtendFXApp() throws java.lang.Exception failed: java.lang.reflect.InvocationTargetException 
----begin detailed exceptions----
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at TestHelper.run(TestHelper.java:200)
	at FXLauncherTest.main(FXLauncherTest.java:469)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.Exception: Failed: testExtendFXApp_useCP_noJAC:2
	at FXLauncherTest.checkStatus(FXLauncherTest.java:205)
	at FXLauncherTest.testExtendFXApp(FXLauncherTest.java:369)
	at FXLauncherTest.testExtendFXApp(FXLauncherTest.java:323)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 7 more
----end detailed exceptions----
_____________________________________________
test# 3: abort on missing Main-Class
JavaFX-Application-Class: helloworld.HelloWorld
renaming fxtest.jar to testMissingMC-3-fxtest.jar
FAIL: testMissingMC:3 : test with helloworld.HelloWorld
Test static void FXLauncherTest.testMissingMC() throws java.lang.Exception failed: java.lang.reflect.InvocationTargetException 
----begin detailed exceptions----
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at TestHelper.run(TestHelper.java:200)
	at FXLauncherTest.main(FXLauncherTest.java:469)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.Exception: Failed: testMissingMC:3
	at FXLauncherTest.checkStatus(FXLauncherTest.java:205)
	at FXLauncherTest.testMissingMC(FXLauncherTest.java:399)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 7 more
----end detailed exceptions----
_____________________________________________
test# 4: test for erroneous jfxrt.jar loading
Main-Class: helloworld.HelloJava
renaming fxtest.jar to testExtraneousJars_useCP-4-fxtest.jar
FAIL: testExtraneousJars_useCP:4 : test with helloworld.HelloJava
Test static void FXLauncherTest.testExtraneousJars() throws java.lang.Exception failed: java.lang.reflect.InvocationTargetException 
----begin detailed exceptions----
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at TestHelper.run(TestHelper.java:200)
	at FXLauncherTest.main(FXLauncherTest.java:469)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.Exception: Failed: testExtraneousJars_useCP:4
	at FXLauncherTest.checkStatus(FXLauncherTest.java:205)
	at FXLauncherTest.testExtraneousJars(FXLauncherTest.java:453)
	at FXLauncherTest.testExtraneousJars(FXLauncherTest.java:410)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 7 more
----end detailed exceptions----
Total: Passed: 0, Failed 4
STDERR:
Executed command: /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/jdk/bin/java --module-path=mods --add-modules=javafx.graphics -cp /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/scratch/0/fxtest.jar helloworld.HelloWorld one two 

###TestError###: test did not return 0 exit value
++++Begin Test Info++++
Test Status: FAIL
++++Test Environment++++
  PATH=/bin:/usr/bin:/usr/sbin
  DISPLAY=:0
  TEST_IMAGE_DIR=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/test
  _JVM_DWARF_PATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/symbols
  XMODIFIERS=@im=fcitx
  LC_ALL=C
  CLASSPATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/tools/launcher/FXLauncherTest.d:/home/loongson/private/jdk-ls/test/jdk/tools/launcher:/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/test/lib:/home/loongson/private/jdk-ls/test/lib:/home/loongson/public/jtreg/lib/javatest.jar:/home/loongson/public/jtreg/lib/jtreg.jar
  LANG=zh_CN.UTF-8
  HOME=/home/loongson
++++Test Output++++
  Error occurred during initialization of boot layer
  java.lang.module.FindException: Unable to compute the hash of module javafx.graphics
++++Test Stack Trace++++
java.lang.Throwable: current stack of the test
  TestHelper.doExec(TestHelper.java:476)
  TestHelper.doExec(TestHelper.java:437)
  FXLauncherTest.doFxExec(FXLauncherTest.java:252)
  FXLauncherTest.testBasicFXApp(FXLauncherTest.java:292)
  FXLauncherTest.testBasicFXApp(FXLauncherTest.java:266)
  java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  java.base/java.lang.reflect.Method.invoke(Method.java:578)
  TestHelper.run(TestHelper.java:200)
  FXLauncherTest.main(FXLauncherTest.java:469)
  java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  java.base/java.lang.reflect.Method.invoke(Method.java:578)
  com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
  java.base/java.lang.Thread.run(Thread.java:1589)
++++End of Test Info++++

Executed command: /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/jdk/bin/java --module-path=mods --add-modules=javafx.graphics -cp /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/scratch/0/fxtest.jar helloworld.ExtHello one two 

###TestError###: test did not return 0 exit value
++++Begin Test Info++++
Test Status: FAIL
++++Test Environment++++
  PATH=/bin:/usr/bin:/usr/sbin
  DISPLAY=:0
  TEST_IMAGE_DIR=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/test
  _JVM_DWARF_PATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/symbols
  XMODIFIERS=@im=fcitx
  LC_ALL=C
  CLASSPATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/tools/launcher/FXLauncherTest.d:/home/loongson/private/jdk-ls/test/jdk/tools/launcher:/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/test/lib:/home/loongson/private/jdk-ls/test/lib:/home/loongson/public/jtreg/lib/javatest.jar:/home/loongson/public/jtreg/lib/jtreg.jar
  LANG=zh_CN.UTF-8
  HOME=/home/loongson
++++Test Output++++
  Error occurred during initialization of boot layer
  java.lang.module.FindException: Unable to compute the hash of module javafx.graphics
++++Test Stack Trace++++
java.lang.Throwable: current stack of the test
  TestHelper.doExec(TestHelper.java:476)
  TestHelper.doExec(TestHelper.java:437)
  FXLauncherTest.doFxExec(FXLauncherTest.java:252)
  FXLauncherTest.testExtendFXApp(FXLauncherTest.java:350)
  FXLauncherTest.testExtendFXApp(FXLauncherTest.java:323)
  java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  java.base/java.lang.reflect.Method.invoke(Method.java:578)
  TestHelper.run(TestHelper.java:200)
  FXLauncherTest.main(FXLauncherTest.java:469)
  java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  java.base/java.lang.reflect.Method.invoke(Method.java:578)
  com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
  java.base/java.lang.Thread.run(Thread.java:1589)
++++End of Test Info++++

ERROR: launcher did not abort properly
Executed command: /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/jdk/bin/java --module-path=mods --add-modules=javafx.graphics -jar /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/scratch/0/fxtest.jar one two 

###TestError###: string <no main manifest attribute> not found
++++Begin Test Info++++
Test Status: FAIL
++++Test Environment++++
  PATH=/bin:/usr/bin:/usr/sbin
  DISPLAY=:0
  TEST_IMAGE_DIR=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/test
  _JVM_DWARF_PATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/symbols
  XMODIFIERS=@im=fcitx
  LC_ALL=C
  CLASSPATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/tools/launcher/FXLauncherTest.d:/home/loongson/private/jdk-ls/test/jdk/tools/launcher:/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/test/lib:/home/loongson/private/jdk-ls/test/lib:/home/loongson/public/jtreg/lib/javatest.jar:/home/loongson/public/jtreg/lib/jtreg.jar
  LANG=zh_CN.UTF-8
  HOME=/home/loongson
++++Test Output++++
  Error occurred during initialization of boot layer
  java.lang.module.FindException: Unable to compute the hash of module javafx.graphics
++++Test Stack Trace++++
java.lang.Throwable: current stack of the test
  TestHelper.doExec(TestHelper.java:476)
  TestHelper.doExec(TestHelper.java:437)
  FXLauncherTest.doFxExec(FXLauncherTest.java:252)
  FXLauncherTest.testMissingMC(FXLauncherTest.java:390)
  java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  java.base/java.lang.reflect.Method.invoke(Method.java:578)
  TestHelper.run(TestHelper.java:200)
  FXLauncherTest.main(FXLauncherTest.java:469)
  java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  java.base/java.lang.reflect.Method.invoke(Method.java:578)
  com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
  java.base/java.lang.Thread.run(Thread.java:1589)
++++End of Test Info++++

Executed command: /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/jdk/bin/java --module-path=mods --add-modules=javafx.graphics -verbose:class -cp /home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/scratch/0/fxtest.jar helloworld.HelloJava one two 

###TestError###: test did not return 0 exit value
++++Begin Test Info++++
Test Status: FAIL
++++Test Environment++++
  PATH=/bin:/usr/bin:/usr/sbin
  DISPLAY=:0
  TEST_IMAGE_DIR=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/test
  _JVM_DWARF_PATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/images/symbols
  XMODIFIERS=@im=fcitx
  LC_ALL=C
  CLASSPATH=/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/tools/launcher/FXLauncherTest.d:/home/loongson/private/jdk-ls/test/jdk/tools/launcher:/home/loongson/private/jdk-ls/build/linux-loongarch64-server-fastdebug/test-support/jtreg_test_jdk_tools_launcher_FXLauncherTest_java/classes/0/test/lib:/home/loongson/private/jdk-ls/test/lib:/home/loongson/public/jtreg/lib/javatest.jar:/home/loongson/public/jtreg/lib/jtreg.jar
  LANG=zh_CN.UTF-8
  HOME=/home/loongson
++++Test Output++++
  [0.016s][info][class,load] java.lang.Object source: shared objects file
  [0.019s][info][class,load] java.io.Serializable source: shared objects file
...
```

Thanks,
Leslie Zhai

**Comment by AlanBateman** [comment]
The --patch-module option is used to override or add classes/resources in the module, it doesn't override the module definition. I see you tried --upgrade-module-path too but you've set the value to IMPORT_MODULE_DIR and it's not clear what this means in this test.

Can you try this, I think this should do what you want:

```
--- a/test/jdk/tools/launcher/FXLauncherTest.java
+++ b/test/jdk/tools/launcher/FXLauncherTest.java
@@ -223,7 +223,7 @@ public class FXLauncherTest extends TestHelper {
         compileFXModule();
 
         List<String> fxCompilerArgs = new ArrayList<>();
-        fxCompilerArgs.add("--module-path=" + MODULE_DIR);
+        fxCompilerArgs.add("--upgrade-module-path=" + MODULE_DIR);
         fxCompilerArgs.add("--add-modules=javafx.graphics");
         fxCompilerArgs.addAll(Arrays.asList(compilerArgs));
         compile(fxCompilerArgs.toArray(new String[fxCompilerArgs.size()]));
@@ -232,7 +232,7 @@ public class FXLauncherTest extends TestHelper {
     static TestResult doFxExec(String...cmds) {
         List<String> fxCmds = new ArrayList<>();
         fxCmds.addAll(Arrays.asList(cmds));
-        fxCmds.add(1, "--module-path=" + MODULE_DIR);
+        fxCmds.add(1, "--upgrade-module-path=" + MODULE_DIR);
         fxCmds.add(2, "--add-modules=javafx.graphics");
         return doExec(fxCmds.toArray(new String[fxCmds.size()]));
     }
@@ -439,18 +439,6 @@ public class FXLauncherTest extends TestHelper {
     }
 
     public static void main(String... args) throws Exception {
-
-        // Ensure that FX is not part of jdk
-        Class<?> fxClass = null;
-        try {
-            fxClass = Class.forName(FX_MARKER_CLASS);
-        } catch (ClassNotFoundException ex) {
-            // do nothing
-        }
-        if (fxClass != null) {
-            throw new RuntimeException("JavaFX modules erroneously included in the JDK");
-        }
-
         FXLauncherTest fxt = new FXLauncherTest();
         fxt.run(args);
         if (testExitValue > 0) {

```

**Comment by xiangzhai** [comment]
Hi @AlanBateman 

Debug patch:

```
diff --git a/test/jdk/tools/launcher/FXLauncherTest.java b/test/jdk/tools/launcher/FXLauncherTest.java
index 9cc4b5aea98..33c3e2788ad 100644
--- a/test/jdk/tools/launcher/FXLauncherTest.java
+++ b/test/jdk/tools/launcher/FXLauncherTest.java
@@ -223,7 +223,7 @@ public class FXLauncherTest extends TestHelper {
         compileFXModule();
 
         List<String> fxCompilerArgs = new ArrayList<>();
-        fxCompilerArgs.add("--module-path=" + MODULE_DIR);
+        fxCompilerArgs.add("--upgrade-module-path=" + MODULE_DIR);
         fxCompilerArgs.add("--add-modules=javafx.graphics");
         fxCompilerArgs.addAll(Arrays.asList(compilerArgs));
         compile(fxCompilerArgs.toArray(new String[fxCompilerArgs.size()]));
@@ -232,7 +232,7 @@ public class FXLauncherTest extends TestHelper {
     static TestResult doFxExec(String...cmds) {
         List<String> fxCmds = new ArrayList<>();
         fxCmds.addAll(Arrays.asList(cmds));
-        fxCmds.add(1, "--module-path=" + MODULE_DIR);
+        fxCmds.add(1, "--upgrade-module-path=" + MODULE_DIR);
         fxCmds.add(2, "--add-modules=javafx.graphics");
         return doExec(fxCmds.toArray(new String[fxCmds.size()]));
     }
@@ -439,18 +439,6 @@ public class FXLauncherTest extends TestHelper {
     }
 
     public static void main(String... args) throws Exception {
-
-        // Ensure that FX is not part of jdk
-        Class<?> fxClass = null;
-        try {
-            fxClass = Class.forName(FX_MARKER_CLASS);
-        } catch (ClassNotFoundException ex) {
-            // do nothing
-        }
-        if (fxClass != null) {
-            throw new RuntimeException("JavaFX modules erroneously included in the JDK");
-        }
-
         FXLauncherTest fxt = new FXLauncherTest();
         fxt.run(args);
         if (testExitValue > 0) {
```

Still failed:  [debug.log](https://github.com/openjdk/jdk/files/9614161/debug.log)

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
> but you've set the value to IMPORT_MODULE_DIR and it's not clear what this means in this test

I just want to try the imported javafx modules:

```
--patch-module=javafx.graphics=build/linux-loongarch64-server-fastdebug/images/jdk/jmods/javafx.graphics.jmod
```

**Comment by AlanBateman** [comment]
I've created JDK-8294093 to track changing ReadSingleImportMetaData (in Modules.gmk) to ignore the mapping in build.properties. If nothing else, that will avoid surprises that will otherwise arise if you run jlink to create a run-time image that includes the javafx.* modules; when you run jlink then it does not map unknown modules to the boot or platform class loader so they will be mapped to the application class loader.

The other aspect to this is whether javafx.* modules should be upgradable. The build adds the imported modules PLATFORM_MODULES but not UPGRADEABLE_PLATFORM_MODULES so this means they will be treated as if they are strongly coupled with the platform/JDK modules and hashed, which probably isn't right here.

**Comment by AlanBateman** [comment]
> Still failed: [debug.log](https://github.com/openjdk/jdk/files/9614161/debug.log)

You are running into module hashes now so I think you'll need some of JDK-8294093. In make/common/Modules.gmk ReadSingleImportMetaData can you add

   UPGRADEABLE_PLATFORM_MODULES += $1

This will add the names of the imported modules to the list of upgradeable modules and should mean they are not hashed.

**Comment by xiangzhai** [comment]
Hi @AlanBateman 

> I've created JDK-8294093 to track changing ReadSingleImportMetaData (in Modules.gmk) to ignore the mapping in build.properties

Thanks for your report!

So 8293977: jdk/modules/etc/VerifyModuleDelegation.java fail with jfx could be fixed too.

> --with-import-modules option was used in JDK 9/10 for co-bundling JavaFX with the JDK. It seems this configure option is still in use, by some.

Yes, @dumasun use it :)

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
Hi @AlanBateman 

> You are running into module hashes now so I think you'll need some of JDK-8294093. In make/common/Modules.gmk ReadSingleImportMetaData can you add
> 
> UPGRADEABLE_PLATFORM_MODULES += $1

debug patch?

```
diff --git a/make/common/Modules.gmk b/make/common/Modules.gmk
index 0eb0fb2ddc1..3431384e168 100644
--- a/make/common/Modules.gmk
+++ b/make/common/Modules.gmk
@@ -272,6 +272,7 @@ FindModuleLegalSrcDirs = \
 
 # Param 1 - Name of module
 define ReadSingleImportMetaData
+    UPGRADEABLE_PLATFORM_MODULES += $1
     ifneq ($$(wildcard $(IMPORT_MODULES_MAKE)/$$(strip $1)/build.properties), )
       classloader :=
       include_in_jre :=
```

Quoted from JDK-8294093:

> --with-import-modules option was used in JDK 9/10 for co-bundling JavaFX with the JDK. It seems this configure option is still in use, by some.

BTW, how to release JavaFX and JDK together? or just release them seperately?

Thanks,
Leslie Zhai

**Pull request closed without merging**